rustc: Use LLVM's new saturating float-to-int intrinsics

alexcrichton opened the pull request:
This commit updates rustc, with an applicable LLVM version, to use
LLVM's new `llvm.fpto{u,s}i.sat.*.*` intrinsics to implement saturating
floating-point-to-int conversions. This results in a little bit tighter
codegen for x86/x86_64, but the main purpose of this is to prepare for
upcoming changes to the WebAssembly backend in LLVM where wasm's
saturating float-to-int instructions will now be implemented with these
intrinsics.

This change allows simplifying a good deal of surrounding code, namely
removing a lot of wasm-specific behavior. WebAssembly no longer has any
special-casing of saturating arithmetic instructions and the need for
`fptoint_may_trap` is gone and all handling code for that is now
removed. This means that the only wasm-specific logic is in the
`fpto{s,u}i` instructions which only get used for "out of bounds is
undefined behavior". This does mean that for the WebAssembly target
specifically the Rust compiler will no longer be 100% compatible with
pre-LLVM 12 versions, but it seems like that's unlikely to be relied on
by too many folks.

Note that this change does immediately regress the codegen of saturating
float-to-int casts on WebAssembly due to the specialization of the LLVM
intrinsic not being present in our LLVM fork just yet. I'll be following
up with an LLVM update to pull in those patches, but affects a few other
SIMD things in flight for WebAssembly so I wanted to separate this change.

Eventually the entire `cast_float_to_int` function can be removed when
LLVM 12 is the minimum version, but that will require sinking the
complexity of it into other backends such as Cranelfit.

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
For comparison, [this is the difference](https://gist.github.com/alexcrichton/09153c3c1f95dd738e19204f43781fdb) on x86_64 of the codegen differences.

alexcrichton commented:
Also, on second though, I went ahead and removed as many WebAssembly-specific bits here as I could, so I think that this represents the final state of what this will look like in the long run (until pre-LLVM-12 is dropped). The consequences of this though is that WebAssembly targets on pre-LLVM-12 won't work (because the intrinsic won't be defined) and the WebAssembly target with `+nontrapping-fptoint` will not have as great codegen as before until https://github.com/llvm/llvm-project/commit/5c729750a6d75df4eeb3eaad72e0b4e93ea27c0e is cherry-picked to our LLVM fork. I plan to do this soon once some other SIMD related things have landed.

nagisa commented:
r? @nagisa 

I don't believe @matthewjasper been active lately.

nagisa commented on compiler/rustc_codegen_llvm/src/builder.rs:
```diff
@@ -1419,9 +1375,4 @@ impl Builder<'a, 'll, 'tcx> {
-
-    fn wasm_and_missing_nontrapping_fptoint(&self) -> bool {
-        self.sess().target.arch == "wasm32"
-            && !self.sess().target_features.contains(&sym::nontrapping_dash_fptoint)
```
`nontrapping_dash_fptoint` should be removed from a list of symbols now that it no longer has any uses remaining.

nagisa commented on compiler/rustc_codegen_ssa/src/mir/rvalue.rs:
```diff
@@ -891,134 +890,39 @@ fn cast_float_to_int<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+    // For unsigned types, the above step already yielded int_ty::MIN == 0 if x is NaN.
+    // Therefore we only need to execute this step for signed integer types.
+    if signed {
+        // LLVM has no isNaN predicate, so we use (x == x) instead
```
Preexisting, but there are both `fcmp uno` and `fcmp ord` which yield true if either operand is a NaN and neither operand is NaN. Not if they'd yield any improvement here, so should just keep as is here for now.

nagisa commented on compiler/rustc_codegen_llvm/src/builder.rs:
```diff
@@ -669,81 +670,36 @@ impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
-        // Note that we skip the wasm intrinsics for vector types where `fptoui`
-        // must be used instead.
-        if self.wasm_and_missing_nontrapping_fptoint() {
+        // This intrinsic is only used with non-saturating casts that have UB on
```
Nit: this is not really an intrinsic, but rather a builder method/instruction builder/etc?

From this comment it is no longer clear why it is important that we specifically go through an effort  on wasm to pick a behaviour for what's otherwise an undefined behaviour. It would be good to keep the remark about the intrinsics producing better machine code.

(as a side note: From spelunking it seems that `simd_cast` also uses this (cc @workingjubilee @calebzulawski) which is potentially undesired (similar to the issue fixed in https://github.com/rust-lang/rust/pull/84274). This implementation would also crash in a simd setting on wasm, but that's a future improvement)

calebzulawski replied:
It does seem a little odd to trap on UB, my understanding is that this function is really only for `{f32,f64}::to_int_unchecked` which assumes the caller already checked the bounds.

This doesn't crash on wasm simd (we test the simd version of `to_int_unchecked` on wasm)--it looks like it falls back to the generic llvm instruction which supports vectors

alexcrichton replied:
I've changed "intrinsic" to "method", and I've elaborated on the comment to hopefully help explain why wasm is different here.

This should work for SIMD types since it specifically avoids (a few lines below) any input types that are vectors.

nagisa reviewed (commented):
This broadly speaking seems good to me, but see remarks inline on some sticking points.

r=me once those are resolved.

bjorn3 commented:
> Eventually the entire cast_float_to_int function can be removed when
LLVM 12 is the minimum version, but that will require sinking the
complexity of it into other backends such as Cranelfit.

Cranelift specifically only has a saturating and trapping version of the float-to-int conversion instruction. There is none that has UB or returns a bogus result.

alexcrichton commented:
@bors: r=nagisa

bors commented:
:pushpin: Commit de2a4601abf69d2eae9845eda20ee5ed7bb9f8af has been approved by `nagisa`

<!-- @bors r=nagisa de2a4601abf69d2eae9845eda20ee5ed7bb9f8af -->
<!-- homu: {"type":"Approved","sha":"de2a4601abf69d2eae9845eda20ee5ed7bb9f8af","approver":"nagisa"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/84389#issuecomment-824077668) for pull requests below priority 1000. This pull request will be tested once the tree is reopened.

ehuss commented:
@bors r-

This failed on the `dist-riscv64-linux` job with a seg fault building gimli: https://github.com/rust-lang/rust/pull/84432#issuecomment-824888056  Confirmed by running that job locally.

nagisa commented:
@bors rollup=never

Sounds like a typical LLVM assertion. @alexcrichton you want to look into this or should somebody from wg-llvm?

alexcrichton commented:
I can try taking a look at this  first and see how far I get

alexcrichton commented:
Hm ok this wasn't too hard to isolate, my first [first guess](https://godbolt.org/z/oM5fe5hYx) worked. Changing the target-cpu to `generic-rv64`, as suggested by trunk LLVM, seems to fix the issue. I'll update the riscv target specs here since that seems like it's probably LLVM's intention here?

alexcrichton commented:
Hm no I take that back, I need to keep digging, that's already set...

alexcrichton commented:
Ok I ended up tracking it down to https://bugs.llvm.org/show_bug.cgi?id=50083 - the same reproduction, just needed the `+f` feature. re-r? @nagisa though?

nagisa commented on compiler/rustc_codegen_llvm/src/builder.rs:
```diff
@@ -1420,8 +1388,11 @@ impl Builder<'a, 'll, 'tcx> {
+    fn fptoint_sat_broken_in_llvm(&self) -> bool {
+        match self.tcx.sess.target.arch.as_str() {
+            // FIXME - https://bugs.llvm.org/show_bug.cgi?id=50083
+            "riscv64" => true,
```
Rather than saying this, can you please say `get_version() < (13, 0, 0)`? This will naturally lead us to removing this workaround once we bump our minimum supported version to 13.0 or later.

alexcrichton replied:
Certainly! Updated now.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 35ae752231dd8ea1abd2a1a970c358eda28b8ebd has been approved by `nagisa`

<!-- @bors r=nagisa 35ae752231dd8ea1abd2a1a970c358eda28b8ebd -->
<!-- homu: {"type":"Approved","sha":"35ae752231dd8ea1abd2a1a970c358eda28b8ebd","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 35ae752231dd8ea1abd2a1a970c358eda28b8ebd with merge 1e747ccec99be0974f47d70abed4d94eafa1d4d9...
<!-- homu: {"type":"BuildStarted","head_sha":"35ae752231dd8ea1abd2a1a970c358eda28b8ebd","merge_sha":"1e747ccec99be0974f47d70abed4d94eafa1d4d9"} -->

rust-log-analyzer commented:
The job **`wasm32`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2420874878) [(plain)](https://github.com/rust-lang-ci/rust/commit/1e747ccec99be0974f47d70abed4d94eafa1d4d9/checks/2420874878/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] codegen/wasm_casts_trapping.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll" "/checkout/src/test/codegen/wasm_casts_trapping.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
/checkout/src/test/codegen/wasm_casts_trapping.rs:10:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:7:35: note: scanning from here
define dso_local i64 @cast_f64_i64(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:9:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptosi.sat.i64.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:19:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:14:35: note: scanning from here
define dso_local i32 @cast_f64_i32(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:16:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptosi.sat.i32.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:28:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:21:35: note: scanning from here
define dso_local i64 @cast_f32_i64(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:23:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptosi.sat.i64.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:37:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:28:35: note: scanning from here
define dso_local i32 @cast_f32_i32(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:30:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptosi.sat.i32.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:46:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:35:35: note: scanning from here
define dso_local i64 @cast_f64_u64(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:37:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptoui.sat.i64.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:55:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:42:35: note: scanning from here
define dso_local i32 @cast_f64_u32(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:44:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptoui.sat.i32.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:64:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:49:35: note: scanning from here
define dso_local i64 @cast_f32_u64(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:51:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptoui.sat.i64.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:73:12: error: CHECK: expected string not found in input
 // CHECK: {{.*}} call {{.*}} @llvm.wasm.trunc.{{.*}}
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:56:35: note: scanning from here
define dso_local i32 @cast_f32_u32(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:58:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptoui.sat.i32.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_trapping.rs:81:12: error: CHECK: expected string not found in input
 // CHECK: fptoui float {{.*}} to i8
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:63:33: note: scanning from here
define dso_local i8 @cast_f32_u8(float %a) unnamed_addr #0 {
                                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll:65:40: note: possible intended match here
 %0 = tail call i8 @llvm.fptoui.sat.i8.f32(float %a)

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_trapping/wasm_casts_trapping.ll
Check file: /checkout/src/test/codegen/wasm_casts_trapping.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'wasm_casts_trapping.3a1fbbbh-cgu.0'
            2: source_filename = "wasm_casts_trapping.3a1fbbbh-cgu.0"
            3: target datalayout = "e-m:e-p:32:32-i64:64-n32:64-S128"
            4: target triple = "wasm32-unknown-emscripten"
            5: 
            6: ; Function Attrs: nounwind readnone uwtable willreturn
            7: define dso_local i64 @cast_f64_i64(double %a) unnamed_addr #0 {
check:10'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            8: start:
check:10'0     ~~~~~~
            9:  %0 = tail call i64 @llvm.fptosi.sat.i64.f64(double %a)
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:10'1           ?                                                 possible intended match
           10:  ret i64 %0
check:10'0     ~~~~~~~~~~~
           11: }
check:10'0     ~
           12: 
check:10'0     ~
           13: ; Function Attrs: nounwind readnone uwtable willreturn
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14: define dso_local i32 @cast_f64_i32(double %a) unnamed_addr #0 {
check:10'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:19'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           15: start:
check:19'0     ~~~~~~
           16:  %0 = tail call i32 @llvm.fptosi.sat.i32.f64(double %a)
check:19'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:19'1           ?                                                 possible intended match
           17:  ret i32 %0
check:19'0     ~~~~~~~~~~~
           18: }
check:19'0     ~
           19: 
check:19'0     ~
           20: ; Function Attrs: nounwind readnone uwtable willreturn
check:19'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21: define dso_local i64 @cast_f32_i64(float %a) unnamed_addr #0 {
check:19'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:28'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           22: start:
check:28'0     ~~~~~~
           23:  %0 = tail call i64 @llvm.fptosi.sat.i64.f32(float %a)
check:28'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:28'1           ?                                                possible intended match
           24:  ret i64 %0
check:28'0     ~~~~~~~~~~~
           25: }
check:28'0     ~
           26: 
check:28'0     ~
           27: ; Function Attrs: nounwind readnone uwtable willreturn
check:28'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28: define dso_local i32 @cast_f32_i32(float %a) unnamed_addr #0 {
check:28'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:37'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           29: start:
check:37'0     ~~~~~~
           30:  %0 = tail call i32 @llvm.fptosi.sat.i32.f32(float %a)
check:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:37'1           ?                                                possible intended match
           31:  ret i32 %0
check:37'0     ~~~~~~~~~~~
           32: }
check:37'0     ~
           33: 
check:37'0     ~
           34: ; Function Attrs: nounwind readnone uwtable willreturn
check:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35: define dso_local i64 @cast_f64_u64(double %a) unnamed_addr #0 {
check:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           36: start:
check:46'0     ~~~~~~
           37:  %0 = tail call i64 @llvm.fptoui.sat.i64.f64(double %a)
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:46'1           ?                                                 possible intended match
           38:  ret i64 %0
check:46'0     ~~~~~~~~~~~
           39: }
check:46'0     ~
           40: 
check:46'0     ~
           41: ; Function Attrs: nounwind readnone uwtable willreturn
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           42: define dso_local i32 @cast_f64_u32(double %a) unnamed_addr #0 {
check:46'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:55'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           43: start:
check:55'0     ~~~~~~
           44:  %0 = tail call i32 @llvm.fptoui.sat.i32.f64(double %a)
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:55'1           ?                                                 possible intended match
           45:  ret i32 %0
check:55'0     ~~~~~~~~~~~
           46: }
check:55'0     ~
           47: 
check:55'0     ~
           48: ; Function Attrs: nounwind readnone uwtable willreturn
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49: define dso_local i64 @cast_f32_u64(float %a) unnamed_addr #0 {
check:55'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:64'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           50: start:
check:64'0     ~~~~~~
           51:  %0 = tail call i64 @llvm.fptoui.sat.i64.f32(float %a)
check:64'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:64'1           ?                                                possible intended match
           52:  ret i64 %0
check:64'0     ~~~~~~~~~~~
           53: }
check:64'0     ~
           54: 
check:64'0     ~
           55: ; Function Attrs: nounwind readnone uwtable willreturn
check:64'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56: define dso_local i32 @cast_f32_u32(float %a) unnamed_addr #0 {
check:64'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:73'0                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           57: start:
check:73'0     ~~~~~~
           58:  %0 = tail call i32 @llvm.fptoui.sat.i32.f32(float %a)
check:73'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:73'1           ?                                                possible intended match
           59:  ret i32 %0
check:73'0     ~~~~~~~~~~~
           60: }
check:73'0     ~
           61: 
check:73'0     ~
           62: ; Function Attrs: nounwind readnone uwtable willreturn
check:73'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           63: define dso_local i8 @cast_f32_u8(float %a) unnamed_addr #0 {
check:73'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:81'0                                     X~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           64: start:
check:81'0     ~~~~~~
           65:  %0 = tail call i8 @llvm.fptoui.sat.i8.f32(float %a)
check:81'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:81'1                                            ?             possible intended match
           66:  ret i8 %0
check:81'0     ~~~~~~~~~~
           67: }
check:81'0     ~
           68: 
check:81'0     ~
           69: ; Function Attrs: nounwind readnone uwtable willreturn
check:81'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           70: define dso_local i64 @cast_unchecked_f64_i64(double %a) unnamed_addr #0 {
check:81'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
>>>>>>


------------------------------------------


---- [codegen] codegen/wasm_casts_nontrapping.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll" "/checkout/src/test/codegen/wasm_casts_nontrapping.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:8:12: error: CHECK: expected string not found in input
 // CHECK: tail call i64 @llvm.wasm.trunc.saturate.signed.i64.f64(double {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:7:35: note: scanning from here
define dso_local i64 @cast_f64_i64(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:9:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptosi.sat.i64.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:16:12: error: CHECK: expected string not found in input
 // CHECK: tail call i32 @llvm.wasm.trunc.saturate.signed.i32.f64(double {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:14:35: note: scanning from here
define dso_local i32 @cast_f64_i32(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:16:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptosi.sat.i32.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:24:12: error: CHECK: expected string not found in input
 // CHECK: tail call i64 @llvm.wasm.trunc.saturate.signed.i64.f32(float {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:21:35: note: scanning from here
define dso_local i64 @cast_f32_i64(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:23:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptosi.sat.i64.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:32:12: error: CHECK: expected string not found in input
 // CHECK: tail call i32 @llvm.wasm.trunc.saturate.signed.i32.f32(float {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:28:35: note: scanning from here
define dso_local i32 @cast_f32_i32(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:30:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptosi.sat.i32.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:41:12: error: CHECK: expected string not found in input
 // CHECK: tail call i64 @llvm.wasm.trunc.saturate.unsigned.i64.f64(double {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:35:35: note: scanning from here
define dso_local i64 @cast_f64_u64(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:37:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptoui.sat.i64.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:49:12: error: CHECK: expected string not found in input
 // CHECK: tail call i32 @llvm.wasm.trunc.saturate.unsigned.i32.f64(double {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:42:35: note: scanning from here
define dso_local i32 @cast_f64_u32(double %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:44:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptoui.sat.i32.f64(double %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:57:12: error: CHECK: expected string not found in input
 // CHECK: tail call i64 @llvm.wasm.trunc.saturate.unsigned.i64.f32(float {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:49:35: note: scanning from here
define dso_local i64 @cast_f32_u64(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:51:7: note: possible intended match here
 %0 = tail call i64 @llvm.fptoui.sat.i64.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:65:12: error: CHECK: expected string not found in input
 // CHECK: tail call i32 @llvm.wasm.trunc.saturate.unsigned.i32.f32(float {{.*}})
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:56:35: note: scanning from here
define dso_local i32 @cast_f32_u32(float %a) unnamed_addr #0 {
                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:58:7: note: possible intended match here
 %0 = tail call i32 @llvm.fptoui.sat.i32.f32(float %a)
      ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:74:12: error: CHECK: expected string not found in input
 // CHECK: fptoui float {{.*}} to i8
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:63:33: note: scanning from here
define dso_local i8 @cast_f32_u8(float %a) unnamed_addr #0 {
                                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:65:40: note: possible intended match here
 %0 = tail call i8 @llvm.fptoui.sat.i8.f32(float %a)
                                       ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:86:12: error: CHECK: expected string not found in input
 // CHECK: fptosi double {{.*}} to i64
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:70:45: note: scanning from here
define dso_local i64 @cast_unchecked_f64_i64(double %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:72:49: note: possible intended match here
 %0 = tail call i64 @llvm.wasm.trunc.signed.i64.f64(double %a) #4
                                                ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:95:12: error: CHECK: expected string not found in input
 // CHECK: fptosi double {{.*}} to i32
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:77:45: note: scanning from here
define dso_local i32 @cast_unchecked_f64_i32(double %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:79:49: note: possible intended match here
 %0 = tail call i32 @llvm.wasm.trunc.signed.i32.f64(double %a) #4
                                                ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:104:12: error: CHECK: expected string not found in input
 // CHECK: fptosi float {{.*}} to i64
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:84:45: note: scanning from here
define dso_local i64 @cast_unchecked_f32_i64(float %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:86:49: note: possible intended match here
 %0 = tail call i64 @llvm.wasm.trunc.signed.i64.f32(float %a) #4
                                                ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:113:12: error: CHECK: expected string not found in input
 // CHECK: fptosi float {{.*}} to i32
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:91:45: note: scanning from here
define dso_local i32 @cast_unchecked_f32_i32(float %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:93:49: note: possible intended match here
 %0 = tail call i32 @llvm.wasm.trunc.signed.i32.f32(float %a) #4
                                                ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:123:12: error: CHECK: expected string not found in input
 // CHECK: fptoui double {{.*}} to i64
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:98:45: note: scanning from here
define dso_local i64 @cast_unchecked_f64_u64(double %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:100:51: note: possible intended match here
 %0 = tail call i64 @llvm.wasm.trunc.unsigned.i64.f64(double %a) #4
                                                  ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:132:12: error: CHECK: expected string not found in input
 // CHECK: fptoui double {{.*}} to i32
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:105:45: note: scanning from here
define dso_local i32 @cast_unchecked_f64_u32(double %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:107:51: note: possible intended match here
 %0 = tail call i32 @llvm.wasm.trunc.unsigned.i32.f64(double %a) #4
                                                  ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:141:12: error: CHECK: expected string not found in input
 // CHECK: fptoui float {{.*}} to i64
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:112:45: note: scanning from here
define dso_local i64 @cast_unchecked_f32_u64(float %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:114:51: note: possible intended match here
 %0 = tail call i64 @llvm.wasm.trunc.unsigned.i64.f32(float %a) #4
                                                  ^
/checkout/src/test/codegen/wasm_casts_nontrapping.rs:150:12: error: CHECK: expected string not found in input
 // CHECK: fptoui float {{.*}} to i32
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:119:45: note: scanning from here
define dso_local i32 @cast_unchecked_f32_u32(float %a) unnamed_addr #0 {
                                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll:121:51: note: possible intended match here
 %0 = tail call i32 @llvm.wasm.trunc.unsigned.i32.f32(float %a) #4

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/wasm_casts_nontrapping/wasm_casts_nontrapping.ll
Check file: /checkout/src/test/codegen/wasm_casts_nontrapping.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
             1: ; ModuleID = 'wasm_casts_nontrapping.3a1fbbbh-cgu.0'
             2: source_filename = "wasm_casts_nontrapping.3a1fbbbh-cgu.0"
             3: target datalayout = "e-m:e-p:32:32-i64:64-n32:64-S128"
             4: target triple = "wasm32-unknown-emscripten"
             5: 
             6: ; Function Attrs: nounwind readnone uwtable willreturn
             7: define dso_local i64 @cast_f64_i64(double %a) unnamed_addr #0 {
check:8'0                                         X~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
             8: start:
check:8'0       ~~~~~~
             9:  %0 = tail call i64 @llvm.fptosi.sat.i64.f64(double %a)
check:8'0       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:8'1             ?                                                 possible intended match
            10:  ret i64 %0
check:8'0       ~~~~~~~~~~~
            11: }
check:8'0       ~
            12: 
check:8'0       ~
            13: ; Function Attrs: nounwind readnone uwtable willreturn
check:8'0       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            14: define dso_local i32 @cast_f64_i32(double %a) unnamed_addr #0 {
check:8'0       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'0                                        X~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            15: start:
check:16'0      ~~~~~~
            16:  %0 = tail call i32 @llvm.fptosi.sat.i32.f64(double %a)
check:16'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'1            ?                                                 possible intended match
            17:  ret i32 %0
check:16'0      ~~~~~~~~~~~
            18: }
check:16'0      ~
            19: 
check:16'0      ~
            20: ; Function Attrs: nounwind readnone uwtable willreturn
check:16'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            21: define dso_local i64 @cast_f32_i64(float %a) unnamed_addr #0 {
check:16'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:24'0                                        X~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            22: start:
check:24'0      ~~~~~~
            23:  %0 = tail call i64 @llvm.fptosi.sat.i64.f32(float %a)
check:24'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:24'1            ?                                                possible intended match
            24:  ret i64 %0
check:24'0      ~~~~~~~~~~~
            25: }
check:24'0      ~
            26: 
check:24'0      ~
            27: ; Function Attrs: nounwind readnone uwtable willreturn
check:24'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            28: define dso_local i32 @cast_f32_i32(float %a) unnamed_addr #0 {
check:24'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:32'0                                        X~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
            29: start:
check:32'0      ~~~~~~
            30:  %0 = tail call i32 @llvm.fptosi.sat.i32.f32(float %a)
check:32'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:32'1            ?                                                possible intended match
            31:  ret i32 %0
check:32'0      ~~~~~~~~~~~
            32: }
check:32'0      ~
            33: 
check:32'0      ~
---
test result: FAILED. 194 passed; 2 failed; 75 ignored; 0 measured; 0 filtered out; finished in 1.49s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/wasm32-unknown-emscripten/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/codegen" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen" "--stage-id" "stage2-wasm32-unknown-emscripten" "--suite" "codegen" "--mode" "codegen" "--target" "wasm32-unknown-emscripten" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--nodejs" "/emsdk-portable/node/latest/bin/node" "--npm" "/emsdk-portable/node/latest/bin/npm" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/wasm32-unknown-emscripten/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--llvm-version" "12.0.0-rust-1.53.0-nightly" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xray" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 --host= --target wasm32-unknown-emscripten --exclude library/alloc
Build completed unsuccessfully in 0:25:47
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2421115881)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/2421115881","builder_name":"checks-actions"} -->

alexcrichton commented:
I updated the assertions in `wasm_casts_trapping.rs` and renamed the file to `wasm_float_to_int_casts.rs`. I then deleted `wasm_casts_nontrapping.rs` since rustc no longer has any special handling of that feature and the test would just be the same.

@bors: r=nagisa

bors commented:
:pushpin: Commit ed6dd40b28223f16aaabd88e66a26d2aa5cc4a50 has been approved by `nagisa`

<!-- @bors r=nagisa ed6dd40b28223f16aaabd88e66a26d2aa5cc4a50 -->
<!-- homu: {"type":"Approved","sha":"ed6dd40b28223f16aaabd88e66a26d2aa5cc4a50","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit ed6dd40b28223f16aaabd88e66a26d2aa5cc4a50 with merge 481ba16439299eb07058a7107352c558fdba7f96...
<!-- homu: {"type":"BuildStarted","head_sha":"ed6dd40b28223f16aaabd88e66a26d2aa5cc4a50","merge_sha":"481ba16439299eb07058a7107352c558fdba7f96"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2423269436)
Approved by: nagisa
Pushing 481ba16439299eb07058a7107352c558fdba7f96 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2423269436"},"merge_sha":"481ba16439299eb07058a7107352c558fdba7f96"} -->
